guvnor-project/guvnor-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/ResourceResolver.java

```diff
@@ -0,0 +1,574 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.guvnor.common.services.project.backend.server;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.inject.Named;
+
+import org.guvnor.common.services.backend.exceptions.ExceptionUtilities;
+import org.guvnor.common.services.backend.file.LinkedDirectoryFilter;
+import org.guvnor.common.services.backend.file.LinkedDotFileFilter;
+import org.guvnor.common.services.backend.file.LinkedMetaInfFolderFilter;
+import org.guvnor.common.services.backend.util.CommentedOptionFactory;
+import org.guvnor.common.services.project.model.POM;
+import org.guvnor.common.services.project.model.Package;
+import org.guvnor.common.services.project.model.Project;
+import org.guvnor.common.services.project.service.POMService;
+import org.guvnor.common.services.project.service.PackageAlreadyExistsException;
+import org.guvnor.common.services.project.service.ProjectResourceResolver;
+import org.guvnor.structure.backend.backcompat.BackwardCompatibleUtil;
+import org.guvnor.structure.server.config.ConfigGroup;
+import org.guvnor.structure.server.config.ConfigItem;
+import org.guvnor.structure.server.config.ConfigType;
+import org.guvnor.structure.server.config.ConfigurationService;
+import org.uberfire.backend.server.util.Paths;
+import org.uberfire.backend.vfs.Path;
+import org.uberfire.io.IOService;
+import org.uberfire.java.nio.file.DirectoryStream;
+import org.uberfire.java.nio.file.FileSystem;
+import org.uberfire.java.nio.file.Files;
+
+import static org.guvnor.common.services.project.backend.server.ProjectResourcePaths.*;
+
+public abstract class ResourceResolver<T extends Project>
+        implements ProjectResourceResolver<T> {
+
+    protected IOService              ioService;
+    protected POMService             pomService;
+    protected ConfigurationService   configurationService;
+    private   CommentedOptionFactory commentedOptionFactory;
+    protected BackwardCompatibleUtil backward;
+
+    public ResourceResolver() {
+    }
+
+    public ResourceResolver( final @Named( "ioStrategy" ) IOService ioService,
+                             final POMService pomService,
+                             final ConfigurationService configurationService,
+                             final CommentedOptionFactory commentedOptionFactory,
+                             final BackwardCompatibleUtil backward ) {
+        this.ioService = ioService;
+        this.pomService = pomService;
+        this.configurationService = configurationService;
+        this.commentedOptionFactory = commentedOptionFactory;
+        this.backward = backward;
+    }
+
+    public Package doNewPackage( final Package parentPackage,
```

Comment:
"doNewPackage" isn't a really a nice method name; simply "newPackage" would probably be better (yes, I know this as copied and pasted... but it moves from protected to public)